domain/schema/model.go

```diff
@@ -124,6 +124,24 @@ func ModelDDL() *schema.Schema {
 	patches = append(patches,
 		triggersForImmutableTable("model", "", "model table is immutable"),
 
+		// The charm is unmodifiable.
+		// There is a lot of assumptions in the code that the charm is immutable
+		// from modification. This is a safety net to ensure that the charm is
+		// not modified.
+		triggersForUnmodifiableTable("charm_action", "charm action table is immutable"),
+		triggersForUnmodifiableTable("charm_config", "charm config table is immutable"),
+		triggersForUnmodifiableTable("charm_container_mount", "charm container mount table is immutable"),
+		triggersForUnmodifiableTable("charm_container", "charm container table is immutable"),
+		triggersForUnmodifiableTable("charm_device", "charm device table is immutable"),
+		triggersForUnmodifiableTable("charm_extra_binding", "charm extra binding table is immutable"),
+		triggersForUnmodifiableTable("charm_hash", "charm hash table is immutable"),
+		triggersForUnmodifiableTable("charm_manifest_base", "charm manifest base table is immutable"),
+		triggersForUnmodifiableTable("charm_metadata", "charm metadata table is immutable"),
+		triggersForUnmodifiableTable("charm_relation", "charm relation table is immutable"),
+		triggersForUnmodifiableTable("charm_resource", "charm resource table is immutable"),
+		triggersForUnmodifiableTable("charm_storage", "charm storage table is immutable"),
+		triggersForUnmodifiableTable("charm_term", "charm term table is immutable"),
```

Comment:
From this, it isn't obvious why some tables have been included, and why some haven't (`charm` itself, for instance. `charm_download_info`, `charm_category`, etc. as well).

A little more context in the top comment might be useful